### PR TITLE
Apply davidopps comments to TaintController PR

### LINKED ISSUE #41318: Events in NodeController are not working
I noticed that we have no events from NodeController for some time now and I suspect this PR as this touches Event machinery. I see logs like:
```
Could not construct reference to: '&v1.ObjectReference{Kind:"Node", Namespace:"", Name:"gke-gke-large-deploy-default-pool-1-db034145-l952", UID:"a1c9fe9d-f1ce-11e6-ad7a-42010af00002", APIVersion:"", ResourceVersion:"", FieldPath:""}' due to: 'object does not implement the List interfaces'. Will not report event: 'Normal' 'RegisteredNode' 'Node gke-gke-large-deploy-default-pool-1-db034145-l952 event: Registered Node gke-gke-large-deploy-default-pool-1-db034145-l952 in NodeController'
```
I fixed the problem (using types from `pkg/api/v1` instead of `client-go/pkg/api/v1` when creating Event messages), but once I did this unit tests started to panic:
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x12c3983]

goroutine 14 [running]:
panic(0x22ad7c0, 0xc420012150)
	/usr/local/google/home/gmarek/.gvm/gos/go1.7.3/src/runtime/panic.go:500 +0x1ae
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x17d
panic(0x22ad7c0, 0xc420012150)
	/usr/local/google/home/gmarek/.gvm/gos/go1.7.3/src/runtime/panic.go:458 +0x271
k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*RESTClient).Verb(0x0, 0x248ac0c, 0x4, 0xc42011f501)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/client.go:222 +0x63
k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*RESTClient).Post(0x0, 0x1f8)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/client.go:232 +0x4e
k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*events).CreateWithEventNamespace(0xc42019a600, 0xc420752200, 0xed033a0c6, 0x6, 0x43dd37)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go:53 +0xfe
k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*EventSinkImpl).Create(0xc4204f2460, 0xc420752200, 0xc42047a000, 0x30, 0x28)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go:155 +0x60
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.recordEvent(0x31a4160, 0xc4204f2460, 0xc420752200, 0x0, 0x0, 0x0, 0x405c00, 0xc42019a6a0, 0x2493bd7)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:180 +0xec
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.recordToSink(0x31a4160, 0xc4204f2460, 0xc4200fa000, 0xc42019a6a0, 0xc42019a640, 0x2540be400)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:138 +0x1c8
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartRecordingToSink.func1(0xc4200fa000)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:120 +0xb7
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartEventWatcher.func1(0x3199f60, 0xc4204dc7e0, 0xc4204dc7b0)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:238 +0x127
created by k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartEventWatcher
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:240 +0x91
```

cc @deads2k @wojtek-t @piosz

### PR DESCRIPTION
**Pull request by gmarek**
Fix #41318

cc @timothysc @kevin-wangzefeng

### CODE REVIEW

**Comment by gmarek** [comment]
@davidopp - when applying your comments I found a regression in event handling logic. I included the fix in this PR, but I can split it out if you want.

**Review by timothysc — COMMENTED** [review]
/cc @aveshagarwal

File: pkg/controller/node/taint_controller.go [hunk 1]
```diff
@@ -178,20 +191,34 @@ func getMinTolerationTime(tolerations []v1.Toleration) time.Duration {
...
 func NewNoExecuteTaintManager(c clientset.Interface) *NoExecuteTaintManager {
-	return &NoExecuteTaintManager{
-		client:             c,
-		taintEvictionQueue: CreateWorkerQueue(deletePodHandler(c)),
-		taintedNodes:       make(map[string][]v1.Taint),
-		nodeUpdateChannel:  make(chan *nodeUpdateItem, nodeUpdateChannelSize),
-		podUpdateChannel:   make(chan *podUpdateItem, podUpdateChannelSize),
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(api.Scheme, clientv1.EventSource{Component: "controllermanager"})
+	eventBroadcaster.StartLogging(glog.Infof)
```
Comment (gmarek) [inline pkg/controller/node/taint_controller.go]: I copied the logic from NodeController.

File: pkg/controller/node/cidr_allocator.go [hunk 2]
```diff
@@ -79,6 +80,12 @@ func NewCIDRRangeAllocator(client clientset.Interface, clusterCIDR *net.IPNet, s
 	eventBroadcaster := record.NewBroadcaster()
 	recorder := eventBroadcaster.NewRecorder(api.Scheme, clientv1.EventSource{Component: "cidrAllocator"})
 	eventBroadcaster.StartLogging(glog.Infof)
+	if client != nil {
+		glog.V(0).Infof("Sending events to api server.")
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(client.Core().RESTClient()).Events("")})
+	} else {
+		glog.V(0).Infof("No api server defined - no events will be sent to API server.")
```
Comment (davidopp) [inline pkg/controller/node/cidr_allocator.go]: yeah I'm kinda curious too....

File: pkg/controller/node/taint_controller.go [hunk 3]
```diff
@@ -156,19 +164,24 @@ func getPodsAssignedToNode(c clientset.Interface, nodeName string) ([]v1.Pod, er
...
-				tolerationSeconds := *(tolerations[i].TolerationSeconds)
-				if tolerationSeconds < minTolerationTime {
-					if tolerationSeconds < 0 {
-						minTolerationTime = 0
-					} else {
-						minTolerationTime = tolerationSeconds
-					}
-				}
+			tolerationSeconds := *(tolerations[i].TolerationSeconds)
+			if tolerationSeconds < 0 {
```
Comment (davidopp) [inline pkg/controller/node/taint_controller.go]: I guess this could be <= 0

**Review by davidopp — COMMENTED** [review]
I didn't fully review, but just a couple of comments so far.

BTW when I suggested to add more tests, I was thinking e2e, not unit tests... I guess you are implying that there is not much extra benefit to doing it as e2e test instead of unit test?

File: pkg/controller/node/taint_controller.go (hunk 3 above)
Comment (gmarek) [inline pkg/controller/node/taint_controller.go]: Done.

**Comment by gmarek** [comment]
As for tests - e2e tests for that may be quite fragile (e.g. kubelet might be too busy to respond in time, some pods might schedule if run after disruptive tests, etc.). We spent A LOT of time stabilizing scheduler-predicates tests, and in discussions with @dchen1107 we often said that we prefer unit/integration tests for such things.

File: pkg/controller/node/taint_controller.go [hunk 4]
```diff
@@ -430,3 +468,12 @@ func (tc *NoExecuteTaintManager) handleNodeUpdate(nodeUpdate *nodeUpdateItem) {
 		tc.processPodOnNode(podNamespacedName, node.Name, tolerations, taints, now)
 	}
 }
+
+func (tc *NoExecuteTaintManager) emitPodDeletionEvent(nsName types.NamespacedName) {
```
Comment (davidopp) [inline pkg/controller/node/taint_controller.go]: I think I had also suggested an event when the deletion is canceled (e.g. taint is removed before toleration has expired) -- do you think that would be reasonable?

Comment (gmarek) [inline pkg/controller/node/taint_controller.go]: Yes, sorry - it slipped my mind.

**Comment by davidopp** [comment]
ref/ #40355

**Comment by davidopp** [comment]
In #40355 I had suggested adding three tests

1. three pods on the same node
one tolerates T without tolerationSeconds
one tolerates T for some tolerationSeconds==t
one does not tolerate T
Then add taint T to the node and verify that the first one is never evicted, second is evicted after t, and third is evicted immediately
2. same pod configuration as above, but add two taints T and S, and verify all three pods are evicted immediately
3. a pod that tolerates two taints, with different tolerationSeconds, then add both taints and verify that the pod is evicted after the min of the tolerationSeconds

It looks like you did something like 2 and 3 but not 1?

**Comment by gmarek** [comment]
Isn't "Pods with different toleration times are evicted appropriately" test what you wanted?

I didn't do both direction checking so I added it now. PTAL if that's enough.

File: pkg/controller/node/taint_controller_test.go [hunk 5]
```diff
@@ -464,12 +502,57 @@ func TestUpdateNode(t *testing.T) {
 			expectDelete:    false,
 			additionalSleep: 1500 * time.Millisecond,
 		},
+		{
+			description: "Pod with multiple tolerations are victed when first one runs out",
```
Comment (gmarek) [inline pkg/controller/node/taint_controller_test.go]: Done.

File: pkg/controller/node/taint_controller_test.go [hunk 6]
```diff
@@ -464,12 +502,57 @@ func TestUpdateNode(t *testing.T) {
...
+								Type:   v1.PodReady,
+								Status: v1.ConditionTrue,
+							},
+						},
+					},
+				},
+			},
+			oldNode:         testutil.NewNode("node1"),
+			newNode:         addTaintsToNode(testutil.NewNode("node1"), "testTaint1", "taint1", []int{1, 2}),
+			expectDelete:    true,
```
Comment (gmarek) [inline pkg/controller/node/taint_controller_test.go]: I don't understand. We sleep for 1.7s here, so there's no chance of noticing eviction after 100s.

**Comment by davidopp** [comment]
LGTM

We can/should probably do more extensive testing, but this is OK.

**Pull request merged**